File: Remote/Source/Remote Elements/REView/BatteryStatusButtonView.m
//
// BatteryStatusButtonView.m
// Remote
//
// Created by Jason Cardwell on 5/24/11.
// Copyright 2011 Moondeer Studios. All rights reserved.
//
#import "RemoteElementView_Private.h"
#import "ImageView.h"

@implementation BatteryStatusButtonView

- (void)initializeIVARs
{
    [super initializeIVARs];

    if (TARGET_IPHONE_SIMULATOR)
    {
        _batteryLevel = 0.75;
        _batteryState = UIDeviceBatteryStateCharging;
    }
    else
    {
        _batteryLevel = CurrentDevice.batteryLevel;
        _batteryState = CurrentDevice.batteryState;
    }

    [CurrentDevice setBatteryMonitoringEnabled:YES];

    __weak BatteryStatusButtonView * weakself = self;
    [NotificationCenter addObserverForName:UIDeviceBatteryLevelDidChangeNotification
                                    object:CurrentDevice
                                     queue:MainQueue
                                usingBlock:^(NSNotification * note) {
                                    _batteryLevel = [CurrentDevice batteryLevel];
                                    [weakself setNeedsDisplay];
                                }];

    [NotificationCenter addObserverForName:UIDeviceBatteryStateDidChangeNotification
                                    object:CurrentDevice
                                     queue:MainQueue
                                usingBlock:^(NSNotification * note) {
                                    _batteryState = [CurrentDevice batteryState];
                                    [weakself setNeedsDisplay];
                                }];
}

- (void)initializeViewFromModel
{
    [super initializeViewFromModel];

    _frameColor = self.model.icons[UIControlStateNormal].color;
    _plugColor = self.model.icons[UIControlStateSelected].color;
    _lightningColor = self.model.icons[UIControlStateDisabled].color;
    _fillColor = self.model.icons[UIControlStateHighlighted].color;
    _frameIcon = self.model.icons[UIControlStateNormal].colorImage;
    _plugIcon = self.model.icons[UIControlStateSelected].colorImage;
    _lightningIcon = self.model.icons[UIControlStateDisabled].colorImage;
}

- (void)dealloc { [NotificationCenter removeObserver:self]; }

- (CGSize)intrinsicContentSize
{
    UIImage * iconImage = self.model.icon.colorImage;
    return (iconImage ? iconImage.size : REMinimumSize);
}

/**
 * Overrides the `ButtonView` implementation to perform custom drawing of the 'battery' frame,
 * the fill color that indicates battery level, and the icon that indicates battery state.
 */
- (void)drawContentInContext:(CGContextRef)ctx inRect:(CGRect)rect
{
    if (_batteryLevel == -1)
    {
        _batteryLevel = CurrentDevice.batteryLevel;
        _batteryState = CurrentDevice.batteryState;
    }

    CGRect   insetRect = CGRectInset(rect, 2.0f, 2.0f);
    CGSize   frameIconSize = _frameIcon.size;
    CGSize   frameSize = (CGSizeContainsSize(insetRect.size, frameIconSize)
                          ? frameIconSize
                          : CGSizeAspectMappedToSize(frameIconSize, insetRect.size, YES));
    CGRect   frameRect = CGRectMake(CGRectGetMidX(insetRect) - frameSize.width / 2.0,
                                    CGRectGetMidY(insetRect) - frameSize.height / 2.0,
                                    frameSize.width,
                                    frameSize.height);

    [_frameIcon drawInRect:frameRect];

    CGFloat   padding     = frameSize.width * 0.06;
    CGSize    paintSize   = CGSizeMake(frameSize.width - 4 * padding,
                                       frameSize.height - 3 * padding);
    CGRect    paintRect   = CGRectMake(frameRect.origin.x + padding,
                                       frameRect.origin.y + 1.5 * padding,
                                       paintSize.width,
                                       paintSize.height);

    paintRect.size.width *= _batteryLevel;

    UIBezierPath * path = [UIBezierPath bezierPathWithRect:paintRect];

    [self.fillColor setFill];
    [path fill];

    if (_batteryState == UIDeviceBatteryStateFull)
    {
        [_plugIcon drawInRect:CGRectInset(frameRect, padding, padding)];

    }
    else if (_batteryState == UIDeviceBatteryStateCharging)
    {
        CGSize lightningIconSize = _lightningIcon.size;
        CGSize   lightningSize = (CGSizeContainsSize(paintSize, lightningIconSize)
                                  ? lightningIconSize
                                  : CGSizeAspectMappedToSize(lightningIconSize, paintSize, YES));
        CGRect   lightningRect = (CGRect){ .size = lightningSize };

        lightningRect.origin.x = CGRectGetMidX(frameRect) - lightningSize.width / 2.0;
        lightningRect.origin.y = CGRectGetMidY(frameRect) - lightningSize.height / 2.0;
        [_lightningIcon drawInRect:lightningRect];
    }
}

@end

File: UI/Source/BatteryStatusButtonView.swift
//
//  BatteryStatusButtonView.swift
//  Remote
//
//  Created by Jason Cardwell on 11/07/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import UIKit
import MoonKit
import DataModel

public final class BatteryStatusButtonView: ButtonView {

	var batteryFrame: ImageView?
	var batteryPlug: ImageView?
	var batteryLightning: ImageView?
	var batteryFill: ImageView?

  private(set) var batteryLevel: Float = -1
  private(set) var batteryState: UIDeviceBatteryState = .Unknown
  private var batteryLevelReceptionist: MSNotificationReceptionist!
  private var batteryStateReceptionist: MSNotificationReceptionist!

	/** initializeIVARs */
	override func initializeIVARs() {
		super.initializeIVARs()
		let device = UIDevice.currentDevice()
		batteryLevel = device.batteryLevel
		batteryState = device.batteryState
		device.batteryMonitoringEnabled = true
	}

	/** registerForChangeNotification */
	override func registerForChangeNotification() {
		super.registerForChangeNotification()
		let device = UIDevice.currentDevice()
		let queue = NSOperationQueue.mainQueue()
		batteryLevelReceptionist = MSNotificationReceptionist(
			observer: self,
			forObject: device,
			notificationName: UIDeviceBatteryLevelDidChangeNotification,
			queue: queue,
			handler: {
				(receptionist: MSNotificationReceptionist!) -> Void in
					if let v = receptionist.observer as? BatteryStatusButtonView {
						v.batteryLevel = device.batteryLevel
						v.setNeedsDisplay()
					}
			})
		batteryStateReceptionist = MSNotificationReceptionist(
			observer: self,
			forObject: device,
			notificationName: UIDeviceBatteryStateDidChangeNotification,
			queue: queue,
			handler: {
				(receptionist: MSNotificationReceptionist!) -> Void in
					if let v = receptionist.observer as? BatteryStatusButtonView {
						v.batteryState = device.batteryState
						v.setNeedsDisplay()
					}
			})

	}

	/** initializeViewFromModel */
	override func initializeViewFromModel() {
		super.initializeViewFromModel()
    let icons = button.iconSet
    batteryFrame     = icons?[UIControlState.Normal.rawValue] as? ImageView
    batteryPlug      = icons?[UIControlState.Selected.rawValue] as? ImageView
    batteryLightning = icons?[UIControlState.Disabled.rawValue] as? ImageView
    batteryFill      = icons?[UIControlState.Highlighted.rawValue] as? ImageView
	}

	/**
	intrinsicContentSize

	:returns: CGSize
	*/
  override public func intrinsicContentSize() -> CGSize {
    if let iconSize = button.icon?.image?.size where iconSize > RemoteElementView.MinimumSize { return iconSize }
    else { return RemoteElementView.MinimumSize }
  }


	/**
	drawRect:

	:param: rect CGRect
	*/
	override public func drawRect(rect: CGRect) {

    let baseRect = rect.rectByInsetting(dx: 4, dy: 4).integerRect

	  if batteryLevel == -1 {
	  	let device = UIDevice.currentDevice()
	    batteryLevel = device.batteryLevel
	    batteryState = device.batteryState
	  }

    let baseColor: UIColor
    if let color = button.iconSet?.normal?.color { baseColor = color }
    else { baseColor = Painter.defaultBackgroundColor }

    let hasPower = [.Charging, .Full] ∋ batteryState

    let chargeLevel = CGFloat(batteryLevel)

    Painter.drawBatteryStatus(color: baseColor, hasPower: hasPower, chargeLevel: chargeLevel, frame: baseRect)

/*
	  let insetRect = rect.rectByInsetting(dx: 2.0, dy: 2.0)
	  let frameIconSize = batteryFrame?.image?.size ?? CGSize.zeroSize
	  let frameSize = insetRect.size.contains(frameIconSize) ? frameIconSize : frameIconSize.aspectMappedToSize(insetRect.size, binding: true)
	  let frameRect = CGRect(x: insetRect.midX - frameSize.width / 2.0,
                           y: insetRect.midY - frameSize.height / 2.0,
	                         width: frameSize.width,
	                         height: frameSize.height)

	  batteryFrame?.colorImage?.drawInRect(frameRect)

	  let padding = frameSize.width * 0.06
	  let paintSize = CGSize(width: frameSize.width - 4 * padding, height: frameSize.height - 3 * padding)
	  var paintRect = CGRect(x: frameRect.origin.x + padding, y: frameRect.origin.y + 1.5 * padding,
	                         width: paintSize.width, height: paintSize.height)

	  paintRect.size.width *= CGFloat(batteryLevel)

	  let path = UIBezierPath(rect: paintRect)

	  batteryFill.color?.setFill()
	  path.fill()

	  if batteryState == .Full {
	    batteryPlug.colorImage?.drawInRect(frameRect.rectByInsetting(dx: padding, dy: padding))
	  } else if batteryState == .Charging {
	    let lightningIconSize = batteryLightning.image?.size ?? CGSize.zeroSize
      let lightningSize = paintSize.contains(lightningIconSize) ? lightningIconSize : lightningIconSize.aspectMappedToSize(paintSize, binding: true)
	    let lightningRect = CGRect(origin: CGPoint(x: frameRect.midX - lightningSize.width / 2.0, y: frameRect.midY - lightningSize.height / 2.0), size: lightningSize)
	    batteryLightning.colorImage?.drawInRect(lightningRect)
	  }

*/
  }

}
